### PR TITLE
refactor(velodyne)!: combine Velodyne ROS wrappers into a single node

### PR DESCRIPTION
**Pull request by mojomex**
## PR Type

- Improvement

## Related Links

[TIER IV INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-6302) -- corresponding issue
[TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/x/NgE2tg) -- detailed article

This PR depends on the following PR being merged first:
- #127

This PR fixes the following issues (for Velodyne only):
- #94 
- #109 

## Description

This PR combines the Velodyne Driver, HW Interface and HW Monitor wrappers into one. For more info, refer to
- #127 

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

### CODE REVIEW

**Comment by knzo25** [comment]
Not only for this one, but copyright files are missing.
Did we decide on header blocks or pragmas?

File: nebula_ros/src/velodyne/velodyne_ros_wrapper.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  sensor_cfg_ptr_ = new_config;
+  return Status::OK;
+}
+
+void VelodyneRosWrapper::ReceiveScanMessageCallback(std::unique_ptr<velodyne_msgs::msg::VelodyneScan> scan_msg)
+{
+  if (hw_interface_wrapper_)
+  {
+    RCLCPP_ERROR_THROTTLE(get_logger(), *get_clock(), 1000,
+                          "Ignoring received PandarScan. Launch with launch_hw:=false to enable PandarScan replay.");
```
Comment (knzo25) [inline nebula_ros/src/velodyne/velodyne_ros_wrapper.cpp:153]: PandarScan mention in velodyne

**Comment by mojomex** [comment]
# :green_circle: Self-Evaluation

All tests have been performed on a VLP32 sensor.

:green_circle: Pointcloud output with default parameters
:green_circle: HW Monitor is publishing diagnostics continuously

File: nebula_ros/src/velodyne/velodyne_ros_wrapper.cpp (hunk 1 above)
Comment (mojomex) [reply nebula_ros/src/velodyne/velodyne_ros_wrapper.cpp]: Addressed in [877f90f](https://github.com/tier4/nebula/pull/148/commits/877f90f5f512ca192870b7fa4090a709b7376a8c)

Comment (knzo25) [reply nebula_ros/src/velodyne/velodyne_ros_wrapper.cpp]: @mojomex 
While trivial, please remember that the one that writes the comment resolves the conversation :innocent:

File: nebula_ros/src/velodyne/decoder_wrapper.cpp [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  if (
+    hw_interface_ && (packets_pub_->get_subscription_count() > 0 ||
+                      packets_pub_->get_intra_process_subscription_count() > 0)) {
+    if (current_scan_msg_->packets.size() == 0) {
+      current_scan_msg_->header.stamp = packet_msg->stamp;
+    }
+
+    velodyne_msgs::msg::VelodynePacket velodyne_packet_msg{};
+    velodyne_packet_msg.stamp = packet_msg->stamp;
+    std::copy(packet_msg->data.begin(), packet_msg->data.end(), velodyne_packet_msg.data.begin());
```
Comment (knzo25) [inline nebula_ros/src/velodyne/decoder_wrapper.cpp:169]: Please correct me if I am wrong. Is the copy necessary?

Comment (mojomex) [reply nebula_ros/src/velodyne/decoder_wrapper.cpp:169]: Sadly VelodynePacket uses `std::array` while NebulaPacket uses `std::vector`. If both were `vector`, we could indeed swap them instead but as it is, this copy is necessary.

Comment (knzo25) [reply nebula_ros/src/velodyne/decoder_wrapper.cpp:169]: Thanks!

File: nebula_ros/include/nebula_ros/velodyne/velodyne_ros_wrapper.hpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  /// @brief Get current status of this driver
+  /// @return Current status
+  Status GetStatus();
+
+  /// @brief Start point cloud streaming (Call CloudInterfaceStart of HwInterface)
+  /// @return Resulting status
+  Status StreamStart();
+
+private:
+  void ReceiveCloudPacketCallback(std::vector<uint8_t> & packet);
```
Comment (knzo25) [inline nebula_ros/include/nebula_ros/velodyne/velodyne_ros_wrapper.hpp:54]: nit: several method lack documentation

**Comment by knzo25** [comment]
(I don't know if me being the reviewer is on purpose or it was assigned automatically after made some comments but other that the nit, which can be ignored, LGTM)

File: nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_scan_decoder.hpp [hunk 4]
```diff
@@ -132,23 +133,59 @@ enum RETURN_TYPE {
...
+
+    // Check if scan is complete
+    uint32_t packet_first_azm = packet[2];  // lower word of azimuth block 0
+    packet_first_azm |= packet[3] << 8;     // higher word of azimuth block 0
+
+    uint32_t packet_last_azm = packet[1102];
+    packet_last_azm |= packet[1103] << 8;
+
+    uint32_t packet_first_azm_phased = (36000 + packet_first_azm - phase) % 36000;
+    uint32_t packet_last_azm_phased = (36000 + packet_last_azm - phase) % 36000;
```
Comment (drwnz) [inline nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_scan_decoder.hpp:167]: Not important - but would be nice to define 1102, 1103, 36000 in velodyne_packet.hpp

**Review by drwnz — COMMENTED** [review]
HW monitor needs a bit of tidying but for this PR, LGTM!

File: nebula_ros/include/nebula_ros/velodyne/velodyne_ros_wrapper.hpp (hunk 3 above)
Comment (mojomex) [reply nebula_ros/include/nebula_ros/velodyne/velodyne_ros_wrapper.hpp:54]: I'll address that in a separate PR, a lot of the code needs documenting in other modules as well.

File: nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_scan_decoder.hpp (hunk 4 above)
Comment (mojomex) [reply nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_scan_decoder.hpp]: Addressed in [0064bb8](https://github.com/tier4/nebula/pull/148/commits/0064bb8d6eb129c1b23437eab38db6b25ce2e964)

**Pull request merged**